plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -2521,17 +2521,32 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         HiveTableHandle hiveTable = (HiveTableHandle) table;
 
         List<ColumnHandle> partitionColumns = ImmutableList.copyOf(hiveTable.getPartitionColumns());
-        List<HivePartition> partitions = partitionManager.getOrLoadPartitions(metastore, new HiveIdentity(session), hiveTable);
+        TupleDomain<ColumnHandle> predicate = hiveTable.getEnforcedConstraint();
+        Optional<DiscretePredicates> discretePredicates = Optional.empty();
 
-        TupleDomain<ColumnHandle> predicate = createPredicate(partitionColumns, partitions);
+        // We won't compute the predicate and discretePredicates if the partitions details are not allowed to parse.
+        if (hiveTable.getPartitionNames().isEmpty()) {
+            Optional<List<HivePartition>> partitions = hiveTable.getPartitions();
 
-        Optional<DiscretePredicates> discretePredicates = Optional.empty();
-        if (!partitionColumns.isEmpty()) {
-            // Do not create tuple domains for every partition at the same time!
-            // There can be a huge number of partitions so use an iterable so
-            // all domains do not need to be in memory at the same time.
-            Iterable<TupleDomain<ColumnHandle>> partitionDomains = Iterables.transform(partitions, (hivePartition) -> TupleDomain.fromFixedValues(hivePartition.getKeys()));
-            discretePredicates = Optional.of(new DiscretePredicates(partitionColumns, partitionDomains));
+            if (partitions.isEmpty()) {
+                // Since getTableProperties doesn't allow us to update ConnectorTableHandle, we might throw away this post creating TableProperties.
+                HivePartitionResult partitionResult = partitionManager.getPartitions(metastore, new HiveIdentity(session), table, new Constraint(hiveTable.getEnforcedConstraint()));
+                if (partitionManager.hasPartitionsLessThanThreshold(partitionResult)) {
+                    partitions = Optional.of(partitionManager.getPartitionsAsList(partitionResult));
+                }
+            }
+            if (partitions.isPresent()) {
+                List<HivePartition> hivePartitions = partitions.orElseThrow();
+                predicate = predicate.intersect(createPredicate(partitionColumns, hivePartitions));
```

Comment:
Add a comment why we need to perform intersection. Do we actually need to do it? Can loaded partitions be outside of `getEnforcedConstraint` or is it about something else?